### ISSUE TITLE
Support Point and Line Geometry

### ISSUE THREAD

**Issue by gkjohnson** [gkjohnson/three-mesh-bvh#103]
Support fast raycasting against clouds of points and lines.

Lines are a bit more complicated but points could just be considered every vertex of every triangle. While it wouldn't be completely efficient and some points would be checked redundantly it could still work pretty well.

The big complication here is that points an lines are often consistent in screen space rather than world space, which makes traversing the bounds tree more difficult. I suppose every bound could be extended by the size of a point / line at the closest distance to ensure encapsulation.

_EDIT: It might be possible to make a minimal change that will consider a "range" of points when sorting indices for the bounds. For "triangles" the range or stride is 3 but maybe it could be made variable so points (1) and lines (2) could be used_

**Comment by Gassepouille** [comment]
Hi, 

sorry to bother, but is there any update on this (especially for point cloud )? 

How can I help ?

**Comment by gkjohnson** [comment]
Hello! Unfortunately I have no plans to work on this at the moment. I think adjusting the BVH class to be generally applicable for points or lines would be a pretty big change and I'd want to take some time to think through the most maintainable way to support something like this.

One workaround, though, would be to create a mesh with a three identical triangle indices for each point which would at least get you a well-formed BVH for points. While it wouldn't be optimal it might be the simplest solution and provide big improvements performance-wise.

Of course you'd have to add some custom raycasting logic to support the degenerate triangles but I think the [shapecast](https://github.com/gkjohnson/three-mesh-bvh#shapecast) function should provide enough flexibility to do this. Here are a few cases that would have to be handled:

- When checking the bounds they need to be expanded by the point radius in world space or by a radius based on the camera position and projection (if points are a consistent size in screen space).
- When checking intersection with the "triangle" you'll just need to check the distance to a single vertex to determine intersection (assuming the point is a sphere).

> How can I help ?

I appreciate the offer -- if the above workaround is interesting to you I'd be happy to include a new example demonstrating how to use the library with points if you'd like to work on that. And if it makes sense maybe some of the functionality can be included in the core library. My time is very limited at the moment so I can't put a ton of effort into this myself but I can definitely help out, give reviews, and guidance on how to achieve this.

> by the way, thanks for the work, on mesh it works beautifully and I have a huge perf improvement

Glad to hear it! What are you using the project for if I can ask? I always appreciate hearing how my work is being used.

**Comment by Gassepouille** [comment]
Hey, 

Thanks for the quick answer ! 

Yeah I totally understand, it didn't seem like a trivial change.

I think the workaround might be a good temporary solution for me: that's an excellent idea, I'm try it today!

> My time is very limited at the moment so I can't put a ton of effort into this myself but I can definitely help out, give reviews, and guidance on how to achieve this.

Thanks, I don't want to use too much of your time though, but if I'm block, I'll ping you because I'm definitely not as proficient with this as you are :D 


> Glad to hear it! What are you using the project for if I can ask? I always appreciate hearing how my work is being used.

So I cannot give you a link here because our product is still in "closed beta", but if you're interested, I can send it to you via linkedin. But basically, we have an Iphone application using lidar to generate point clouds and the said application is connected to a web platform to handle those scans. From the point clouds we generate also meshes (that's where I gain a lot in perf) but the main focus is for point clouds.

**Comment by gkjohnson** [comment]
> I think the workaround might be a good temporary solution for me: that's an excellent idea, I'm try it today!

In practice I think the run time performance should be just fine. Build time might be more than it would be with a more optimal solution and index memory will be three times as large as necessary but I don't think that's so bad. I would think it should be a pretty good solution even for the long term. I'll be curious about how it turns out.

> Thanks, I don't want to use too much of your time though, but if I'm block, I'll ping you because I'm definitely not as proficient with this as you are :D

I didn't want to make it sound like I can't help -- I just don't have time to work directly on every feature request myself but I'm happy to help with any direction / questions. As much as possible I want to encourage contributions, though, because ultimately it makes the project better for everyone. I do quite a bit of work with point clouds myself so I may wind up being able to use a solution like this at some point, as well.

The more I think about the more I think something like this would at least make a good example for the project so do feel free to contribute if you can! And I can help improve it when I have time.

> So I cannot give you a link here because our product is still in "closed beta", but if you're interested, I can send it to you via linkedin.

If you can I'd be interested in taking a look!

**Comment by Gassepouille** [comment]
> In practice I think the run time performance should be just fine. Build time might be more than it would be with a more optimal solution and index memory will be three times as large as necessary but I don't think that's so bad. I would think it should be a pretty good solution even for the long term. I'll be curious about how it turns out.

I'm not sure about this: t probably will play well up to a certain number of vertices but I'm afraid it won't play as well above a certain limit ( so far the maximum vertices that we had on a point cloud was 12 000 000). I think we'll see.

> I didn't want to make it sound like I can't help -- I just don't have time to work directly on every feature request myself but I'm happy to help with any direction / questions. As much as possible I want to encourage contributions, though, because ultimately it makes the project better for everyone. I do quite a bit of work with point clouds myself so I may wind up being able to use a solution like this at some point, as well.

No worries,  you didn't come across like this at all! 
Yeah, I couldn't agree more! I'll let you know my finding (I unfortunately didn't have the time to try the implementation yesterday)

 > The more I think about the more I think something like this would at least make a good example for the project so do feel free to contribute if you can! And I can help improve it when I have time.
 
 Sweet, I really appreciate it ! 
 
 > If you can I'd be interested in taking a look! 
 
 I'm sending you the link right now. I'd even love to have your feedback/opinion on it

**Comment by gkjohnson** [comment]
> So, the first "issue" is, since I am now using the THREE.Mesh instead of THREE.Point and the triangles are composed from the same points. My point cloud is not rendered. So I'm thinking:
> - Whether I have 2 models, 1 to render and the other one for the collision

Yes I would create a duplicate mesh that you don't render that has an index which is used for the BVH. The BVH only rearranges the index buffer so you can share all the other buffer attributes to save memory.

> The second "issue" I had, is that I'm not sure how to use the shapecast function ...
> Actually not sure what mouseVector should be since I usually raycast directly from the mouse coordinate.

To most efficiently do the bounds and distance checks you'll want to transform the ray into the local coordinate frame of the point cloud / BVH because all the bounds are axis aligned boxes. You can do that like so:

```js
const inverseMatrix = new THREE.Matrix4();
inverseMatrix.copy( pointCloud.matrixWorld ).invert();

raycaster.setFromCamera( mouse, camera );
raycaster.ray.applyMatrix4( inverseMatrix );

// perform shapecast using raycaster.ray
```

> So the intersectsBoundsFunc :
> ```js
> box => {
>    box.expandByscalar(pointScaleDependingOnCamera);
>    return box.containsPoint( mouseVector);
> }
> ```

I think this is a fine way to start. Because the points are all screenspace size the most rigorous approach would be to find the furthest point on the bounds from the camera and use that to determine the "margin" to add to the box but we can look into that later. I'm not sure how three.js' built in point raycasting works but this would let it be pixel perfect.

> - The intersectsTriangleFunc :
> ```js
> tri=>{
>     tri.closestPointToPoint( mouseVector, targetVector);
>     targetVector.sub( mouseVector );
>     const distance = targetVector.length();
>     if ( distance < pointScaleDependingOnCamera ) {
>         collided = true;
>     }
> }
> ```

For this you can use the [Ray.distanceToPoint](https://threejs.org/docs/?q=ray#api/en/math/Ray.distanceToPoint) function and just test to make sure _one_ of the vertices is within the "intersection" threshold because we know all triangle vertices are at the same point.

Hopefully that helps with some of the road blocks -- let me know if you have other questions. Thanks again for putting time into this!

**Comment by Gassepouille** [comment]
Dude, you're the F***ing man ! 
It's working perfectly! 

Let me do some clean up today, and tomorrow I'll work on an example for you, that's the least I can do!!!

**Comment by Gassepouille** [comment]
I started working on an example, but I'm not too sure how to run the repo locally (npm start gave me some 404/403) 
and how I can create a branch and then a PR for your repo ? :/

By the way, I noticed some precision issue : when I click on one point, it gives me the coordinates of another point -_-'
I'm thinking maybe it has something to do with the scale of the point that I fucked up somehow, I'll look into this today.

**Comment by gkjohnson** [comment]
> I started working on an example, but I'm not too sure how to run the repo locally (npm start gave me some 404/403)

Ah yes sorry this could be documented better. The project uses parcel to build because bare module imports are not supported in the browser. You have to run `npm start` and then visit `localhost:9080/example/raycast.html` or whatever page is you're working on. You should be able to just add a new html page to the example folder and have it be built. I've found parcel to be a bit cumbersome and takes a bit to build but I haven't gotten around to moving to a different build process, yet.

> how I can create a branch and then a PR for your repo ? :/

You can fork the repo for your own user, clone and make a branch in your own repo, and then make a PR from that.

> By the way, I noticed some precision issue : when I click on one point, it gives me the coordinates of another point -_-'
I'm thinking maybe it has something to do with the scale of the point that I fucked up somehow, I'll look into this today.

The scaling of the points can be tricky because they're a consistent size in screen space. With the example I can help out a bit more easily.

**Comment by Gassepouille** [comment]
Hey, I'm so sorry, it seems I'm having some trouble making the examples working locally: 
I did 
- clone repo
- npm install /sudo npm install
- npm start
- go to http://localhost:9080/example/raycast.html
And I keep getting those : 
![screenshot](https://user-images.githubusercontent.com/1702393/118461144-a53ab000-b6fd-11eb-827e-8b3e322c844c.png)

By the way, why not using using the native import ? 

> The scaling of the points can be tricky because they're a consistent size in screen space. With the example I can help out a bit more easily.

Yeah, I managed to fix it, it was  a referencing issue, coming from me not cloning the vector3 I got from the triangle when sending it to wherever I needed.

**Comment by gkjohnson** [comment]
> go to http://localhost:9080/example/raycast.html

Oof sorry I mistyped above. You need to go to go to `http://localhost:9080/example/bundle/raycast.html`. (Note the addition of `/bundle`). Sorry about that.

> By the way, why not using using the native import ?

I wish it were so simple. The library itself uses bare module imports (like `import { ... } from 'three';`) which is required to publish dependencies to npm but still don't work directly in a browser so a build process is needed to resolve those imports. Import maps are intended to solve this but browsers don't support them universally, yet. I agree it's not ideal. I'll have to revisit it at some point.

**Comment by Gassepouille** [comment]
Alright ! 
So I created the example, it's working , here is the PR.
https://github.com/gkjohnson/three-mesh-bvh/pull/239

I tried to get rid of the noise in the PR by running npm build, but no luck :/

As for the parcel alternative, I'm not sure I'm gonna be of any help on this:
For my projects I like to have the least dependencies I can, so no bundler. That being said, for professional projects, I'm usually working with vuejs-cli that have a webpack under the hood. But I think webpack here would be an overkill.
I was looking into this a bit yesterday, and I found this : https://lit.dev/docs/ that could be interesting, but I haven't tried it at all, and the community is not as big as parcel.
